### PR TITLE
[Azure Service Bus] Replace Stopwatch with ValueStopwatch

### PR DESCRIPTION
**Pull request by danielmarbach**
Introduces

https://github.com/dotnet/aspnetcore/blob/master/src/Shared/ValueStopwatch/ValueStopwatch.cs and
https://github.com/dotnet/aspnetcore/blob/master/src/Shared/test/Shared.Tests/ValueStopwatchTest.cs

to avoid allocating Stopwatches all over the place. While taking the file from ASPNETCore I aligned the license with the license of this code. The original license was Apache. I'll let you figure out if that is ok.

### CODE REVIEW

**Comment by pakrym** [comment]
The change looks good, cc @AlexGhiondea for the license question.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Primitives/ValueStopwatch.cs
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Primitives/ValueStopwatch.cs:7]: nit:  We should elevate these before the namespace to match the convention used elsewhere.

**Comment by jsquire** [comment]
Thanks, @danielmarbach!  I'm interested in @AlexGhiondea's decision on this;  I'd like to make use of this in the Event Hubs client as well.    @pakrym:  Assuming that the license change isn't an issue, should we consider moving this to Core?

**Comment by pakrym** [comment]
@jsquire if you are going to use it definitely.

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/Primitives/ValueStopwatchTests.cs
Comment (SeanFeldman) [inline sdk/servicebus/Azure.Messaging.ServiceBus/tests/Primitives/ValueStopwatchTests.cs:37]: Should this test verify that the elapsed time is no less than 200 msec?

Comment (danielmarbach) [reply sdk/servicebus/Azure.Messaging.ServiceBus/tests/Primitives/ValueStopwatchTests.cs:37]: This is the original test from ASPNET Core

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Primitives/ValueStopwatch.cs
Comment (danielmarbach) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Primitives/ValueStopwatch.cs]: Fixed

**Comment by danielmarbach** [comment]
According to my limited research it would also be possible to leave it under Apache 2 in the repo since MIT and Apache 2 are compatible. If that is done the owner doesn't have to be contacted because the license is not changed.

File: sdk/servicebus/Azure.Messaging.ServiceBus/tests/Primitives/ValueStopwatchTests.cs
Comment (SeanFeldman) [reply sdk/servicebus/Azure.Messaging.ServiceBus/tests/Primitives/ValueStopwatchTests.cs:37]: While this is the original test, you could still validate that the time elapsed is no less than the expected value. Otherwise, anything positive would pass w/o necessarily validating that at least the expected time has elapsed.

Comment (danielmarbach) [reply sdk/servicebus/Azure.Messaging.ServiceBus/tests/Primitives/ValueStopwatchTests.cs:37]: I have no objections of doing this once the question with the license is solved

**Comment by danielmarbach** [comment]
Alternatively I can also update the PR to take in https://github.com/dotnet/orleans/blob/master/src/Orleans.Core/Timers/ValueStopwatch.cs which is MIT license already.

Let me know what you think

**Comment by AlexGhiondea** [comment]
@danielmarbach let me look into the license question.

**Comment by danielmarbach** [comment]
Yeah sorry didn't want to cause stress. Was merely thinking out loud about other options. Bad habit of me

**Comment by AlexGhiondea** [comment]
@danielmarbach no worries! Here is what we should do:

In this PR, please use the original copyright header for the file you are copying over.

We’ll have a separate PR to ensure the right attribution happens in the notices.txt file.

**Comment by danielmarbach** [comment]
Should I keep the file as is including the tests or also do the tweaks suggested by @SeanFeldman to the tests?

Also should I already move it to the Core as @jsquire suggested?

**Comment by danielmarbach** [comment]
Brought back the original license and excluded the SA1636 warning in the file. No CI fails with certificate expired failures.

**Comment by jsquire** [comment]
Cool.... looks like we're unblocked for moving forward.    

@pakrym:  Would you prefer that this PR merge as-is and then we do a follow-up to shift things to Core, or should we discuss shifting to Core as part of Daniel's efforts?

**Comment by pakrym** [comment]
Whatever is easier to do. Seems like adding it to core is just a case of moving the source and test files so I don't see a big reason to wait.

**Comment by danielmarbach** [comment]
Who should wait for what? :trollface: 

I wasn't sure based on the above communication what is expected from whom nor the direction to take. So I went ahead and moved the source files to core assuming what I'm doing is somehow aligned with what you have been thinking behind the scenes. 

In general it would be very helpful if external contributors get a bit more guidelines on the PR. Something like:

> Daniel, would you mind moving the file to Core\bla under namespace `foobar` and the edit `bla.props.foobar` to include it as a shared resource? If you don't have time don't worry we can take the further steps before merging the PR.

I hope this doesn't come across in weird ways. I'm reflecting about what would have helped me personally to know what the next steps are from your perspective plus it shortens the communication paths quite a bit. It also gives me the possibility to say "I'm sorry but at this point I've reached the level of how far I can go as a community contributor and I would appreciate if you take it forward from this point"

With ❤️ Daniel

File: sdk/core/Azure.Core/src/Azure.Core.csproj
Comment (pakrym) [inline sdk/core/Azure.Core/src/Azure.Core.csproj:36]: Don't need to include into the core itself, it's not being used there.

**Comment by pakrym** [comment]
Thank you for the feedback!

My bad, I'll do better. You did the right thing here, just one small comment.

File: sdk/core/Azure.Core/src/Azure.Core.csproj
Comment (pakrym) [reply sdk/core/Azure.Core/src/Azure.Core.csproj:36]: To be able to test include shared source file into the test project directly (you'll see examples in the test.csproj)

**Comment by jsquire** [comment]
> Who should wait for what? :trollface:
> 
> I wasn't sure based on the above communication what is expected from whom nor the direction to take. So I went ahead and moved the source files to core assuming what I'm doing is somehow aligned with what you have been thinking behind the scenes.

Apologies;  I'm at fault here.  It wasn't my intent to talk over you with the sidebar and I meant to follow-up yesterday with a more explicit set of details around what we had been discussing.  I got a bit distracted and when I just looped around to do so, it appears that you're already in motion and Pavel was engaged for feedback.

Thank you, again, for taking the time to contribute and to work through the lack of clarity.  Your efforts are very much appreciated.

File: sdk/core/Azure.Core/src/Azure.Core.csproj
Comment (danielmarbach) [reply sdk/core/Azure.Core/src/Azure.Core.csproj]: Addressed *fingerscrossed*

**Comment by danielmarbach** [comment]
@pakrym @jsquire No worries I've been in similar situations on the other side of PRs and have a lot of empathy for things like that. Nonetheless I appreciate that you took the time to listen to the feedback

**Review by jsquire — APPROVED** [review]
LGTM in general and from the Service Bus perspective.   I'll defer to Pavel for the authoritative approval here.

**Review by pakrym — APPROVED** [review]

**Comment by JoshLove-msft** [comment]
@danielmarbach would you be so kind as to rebase this so it can be merged in?

**Comment by danielmarbach** [comment]
@JoshLove-msft done

I also sent in https://github.com/Azure/azure-sdk-for-net/pull/11530 which is a branch of top of this one that also fixes Event Hub to make @jsquire a happy person ;) Once this one is merged I will rebase the other one on top of master

**Comment by jsquire** [comment]
> that also fixes Event Hub to make @jsquire a happy person ;) 

Indeed it does.   Thank you very much for that!

**Pull request merged**
